src/RoomNotifs.js

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2016 OpenMarket Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import MatrixClientPeg from './MatrixClientPeg';
+import PushProcessor from 'matrix-js-sdk/lib/pushprocessor';
+import q from 'q';
+
+export const ALL_MESSAGES_LOUD = 'all_messages_loud';
+export const ALL_MESSAGES = 'all_messages';
+export const MENTIONS_ONLY = 'mentions_only';
+export const MUTE = 'mute';
+
+export function getRoomNotifsState(roomId) {
+    if (MatrixClientPeg.get().isGuest()) return RoomNotifs.ALL_MESSAGES;
+
+    // look through the override rules for a rule affecting this room:
+    // if one exists, it will take precedence.
+    const muteRule = findOverrideMuteRule(roomId);
+    if (muteRule) {
+        return MUTE;
+    }
+
+    // for everything else, look at the room rule.
+    const roomRule = MatrixClientPeg.get().getRoomPushRule('global', roomId);
+
+    // XXX: We have to assume the default is to notify for all messages
+    // (in particular this will be 'wrong' for one to one rooms because
+    // they will notify loudly for all messages)
+    if (!roomRule || !roomRule.enabled) return ALL_MESSAGES;
+
+    // a mute at the room level will still allow mentions
+    // to notify
+    if (isMuteRule(roomRule)) return MENTIONS_ONLY;
+
+    const actionsObject = PushProcessor.actionListToActionsObject(roomRule.actions);
+    if (actionsObject.tweaks.sound) return ALL_MESSAGES_LOUD;
+
+    return null;
+}
+
+export function setRoomNotifsState(roomId, newState) {
+    if (newState == 'mute') {
+        return setRoomNotifsStateMuted(roomId);
+    } else {
+        return setRoomNotifsStateUnmuted(roomId, newState);
+    }
+}
+
+function setRoomNotifsStateMuted(roomId) {
+    const cli = MatrixClientPeg.get();
+    const promises = [];
+
+    // delete the room rule
+    const roomRule = cli.getRoomPushRule('global', roomId);
+    if (roomRule) {
+        promises.push(cli.deletePushRule('global', 'room', roomRule.rule_id));
+    }
+
+    // add an override rule to squelch everything in this room
```

Comment:
suggest rewording this slightly to make it clearer that this will replace existing rules. perhaps just "add/replace override rule ..."